### PR TITLE
[Custom Trainer] Refactor postprocessing

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
## Description

- Move from lightning Trainer to AnomalyTrainer
- Move post-processing from anomaly model to trainer

## Changes

<details>
<summary>Describe the changes you made</summary>

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

</details>

## Checklist

<details>
<summary>Ensure that you followed the following</summary>

- [ ] I have added a summary of my changes to the [CHANGELOG](https://github.com/openvinotoolkit/anomalib/blob/main/CHANGELOG.md) (not for minor changes, docs and tests).
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas (If applicable)
- [ ] I have made corresponding changes to the documentation (If applicable)
- [ ] I have added tests that prove my fix is effective or that my feature works (If applicable)

</details>

### CODE REVIEW

File: src/anomalib/trainer/callbacks/post_processor.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        pl_module: AnomalyModule,
+        outputs: Any,
+        batch: Any,
+        batch_idx: int,
+        dataloader_idx: int = 0,
+    ) -> None:
+        if outputs is not None:
+            self.process(trainer, pl_module, outputs)
+
+    def process(self, trainer: Trainer, pl_module: AnomalyModule, outputs: STEP_OUTPUT):
```
Comment (samet-akcay) [inline src/anomalib/trainer/callbacks/post_processor.py:60]: post_process maybe? Would it be peculiar to name this `self.post_process` when there is also a private method inside, called `self._post_process`?

File: src/anomalib/trainer/trainer.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from .callbacks import PostProcessorCallback
+
+log = logging.getLogger(__name__)
+
+
+class AnomalibTrainer(Trainer):
+    """Anomalib trainer.
+
+    Note:
+        Refer to PyTorch Lightning's Trainer for a list of parameters for details on other Trainer parameters.
```
Comment (samet-akcay) [inline src/anomalib/trainer/trainer.py:22]: Would it be an idea to give a link to the documentation?

File: src/anomalib/trainer/trainer.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        **kwargs,
+    ) -> None:
+        self._setup_callbacks(callbacks)
+        super().__init__(callbacks=callbacks, **kwargs)
+
+        self.lightning_module: AnomalyModule
+
+    def _setup_callbacks(self, callbacks: list[Callback]) -> None:
+        """Setup callbacks for the trainer."""
+        # Note: this needs to be changed when normalization is part of the trainer
```
Comment (samet-akcay) [inline src/anomalib/trainer/trainer.py:40]: Can you also add a comment in the note why this is inserted to idx=0?

**Review by samet-akcay — COMMENTED** [review]
Thanks @ashwinvaidya17, I only have some minor comments.

File: src/anomalib/trainer/trainer.py (hunk 3 above)
Comment (ashwinvaidya17) [reply src/anomalib/trainer/trainer.py:40]: This will be changed later on but currently the rest of the callbacks are passed to the trainer. So post-processing might happen after normalization and thresholding computation is performed. With this, I am ensuring that post-processing is done first.

File: src/anomalib/data/utils/split.py [hunk 4]
```diff
@@ -48,7 +47,7 @@ class ValSplitMode(str, Enum):
     SYNTHETIC = "synthetic"
 
 
-def concatenate_datasets(datasets: Sequence[AnomalibDataset]) -> AnomalibDataset:
+def concatenate_datasets(datasets: Sequence["data.AnomalibDataset"]) -> "data.AnomalibDataset":
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply src/anomalib/data/utils/split.py:50]: If I don't do this, then I get circular import errot. This is because the base dataset imports utils that is in split and Anomalib dataset is referred here.

[the remark this answers is not in the document]
Comment (samet-akcay) [reply src/anomalib/data/utils/split.py:50]: again, a future reference, but would it be possible to do like `anomalib.Dataset` if we have the relative imports properly

**Review by samet-akcay — APPROVED** [review]
Thanks @ashwinvaidya17 !
